src/infra/core/src/query_service_impl.rs

```diff
@@ -357,27 +357,56 @@ impl KamuSchema {
         dataset: &dyn Dataset,
         last_records_to_consider: Option<u64>,
     ) -> Result<Vec<Multihash>, InternalError> {
-        let mut files = Vec::new();
-        let mut num_records = 0;
+        struct DataSliceCollectorVisitorState {
+            files: Vec<Multihash>,
+            num_records: u64,
+            last_records_to_consider: Option<u64>,
+        }
+
+        let mut slice_collector_visitor = GenericCallbackVisitor::new(
+            DataSliceCollectorVisitorState {
+                files: Vec::new(),
+                num_records: 0,
+                last_records_to_consider,
+            },
+            |state, (_, block)| {
```

Comment:
I like this pattern!

I think we can make it even more convenient to use:
```rust
chain.reduce(
  state, // Initial state
  Decision::Next,  // Initial decision
  |state, hash, block| { // Flatten the tuple
    ...
  }
);
```
What do you think?

Reply:
I think it's a cool idea and should be implemented

Reply:
Done